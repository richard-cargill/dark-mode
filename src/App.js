import React from 'react';

const MyContext = React.createContext();

class MyProvider extends React.Component {
  state = {
    current: {
      background: this.props.theme.light.background,
      color: this.props.theme.light.color
    },
    light: {
      background: this.props.theme.light.background,
      color: this.props.theme.light.color
    },
    dark: {
      background: this.props.theme.dark.background,
      color: this.props.theme.dark.color
    }
  }

  render() {
    return (
      <MyContext.Provider value={{
        state: this.state,
        toggleBackground: () => this.setState({
          current: {
            color: this.state.current.color === this.state.light.color ? this.state.dark.color : this.state.light.color,
            background: this.state.current.background === this.state.light.background ? this.state.dark.background : this.state.light.background
          }
        })
      }}>
        {this.props.children}
      </MyContext.Provider>
    )
  }
}

const Family = (props) => (
  <div className="family">
    <ToggleButton className="button" />
  </div>
)

 class Background extends React.Component {
  render() {
    return (
      <MyContext.Consumer>
        {(context) => (
          <div style={{backgroundColor: context.state.current.background, color: context.state.current.color}}>
            {this.props.children}
          </div>
        )}
      </MyContext.Consumer>
    );
  }
}

 class ToggleButton extends React.Component {
  render() {
    return (
      <MyContext.Consumer>
        {(context) => (
          <React.Fragment>
            <button {...this.props} onClick={context.toggleBackground}>Toggle background</button>
          </React.Fragment>
        )}
      </MyContext.Consumer>
    )
  }
}

const theme = {
  light: {
    background: 'black',
    color: 'white'
  },
  dark: {
    background: 'white',
    color: 'black'
  }
};

class App extends React.Component {
  render() {
    return (
      <MyProvider theme={theme}>
        <Background>
          <p>I am the app</p>
          <Family />
        </Background>
      </MyProvider>
    );
  }
}

export {App, ToggleButton, Background}
