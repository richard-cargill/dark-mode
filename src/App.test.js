import React from 'react';
import ReactDOM from 'react-dom';
import {App, ToggleButton, Background} from './App';
import renderer from 'react-test-renderer';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});

test('Div changes colour when button is pressed', () => {
  const component = renderer.create(
    <App />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  const instance = component.root;

  // manually trigger the callback
  const button = instance.find((el) => el.type == 'button');
  button.props.onClick()

  // re-rendering
  tree = component.toJSON();

  const background = instance.find((el) => el.type == 'div');

  expect(background.props.style.backgroundColor).toEqual('white');
  expect(background.props.style.color).toEqual('black');
});
