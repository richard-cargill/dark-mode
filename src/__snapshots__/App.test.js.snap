// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Div changes colour when button is pressed 1`] = `
<div
  style={
    Object {
      "backgroundColor": "black",
      "color": "white",
    }
  }
>
  <p>
    I am the app
  </p>
  <div
    className="family"
  >
    <button
      className="button"
      onClick={[Function]}
    >
      Toggle background
    </button>
  </div>
</div>
`;

exports[`Link changes the class when hovered 1`] = `
<div
  style={
    Object {
      "backgroundColor": "black",
      "color": "white",
    }
  }
>
  <p>
    I am the app
  </p>
  <div
    className="family"
  >
    <button
      className="button"
      onClick={[Function]}
    >
      Toggle background
    </button>
  </div>
</div>
`;

exports[`Link changes the class when hovered 2`] = `
<div
  style={
    Object {
      "backgroundColor": "black",
      "color": "white",
    }
  }
>
  <p>
    I am the app
  </p>
  <div
    className="family"
  >
    <button
      className="button"
      onClick={[Function]}
    >
      Toggle background
    </button>
  </div>
</div>
`;
